File: src/lib/api/ecosystem.ts
import { invoke } from "@tauri-apps/api/core";

export interface Ecosystem {
  id: string;
  name: string;
  description: string;
  isCurrent: boolean;
  createdAt: number;
}

export interface FrameworkRegistry {
  id: string;
  name: string;
  description: string;
  repoUrl: string;
  repoBranch: string;
  installMethod: string;
  filePrefix: string;
  providedDirs: string[];
  isolatedFiles: string[];
}

export interface FrameworkDetail {
  installedAt: number;
  commitHash: string;
}

export interface EcosystemStatus {
  ecoId: string;
  frameworks: string[];
  frameworkDetails: Record<string, FrameworkDetail>;
  mergeConflicts: Record<string, string[]>;
  installErrors: string[];
}

export interface CreateEcosystemResult {
  ecosystem: Ecosystem;
  installErrors: string[];
}

export const ecosystemApi = {
  async list(): Promise<Ecosystem[]> {
    return await invoke("list_ecosystems");
  },

  async getCurrent(): Promise<Ecosystem | null> {
    return await invoke("get_current_ecosystem");
  },

  async create(
    name: string,
    description: string,
    frameworks: string[] = [],
  ): Promise<CreateEcosystemResult> {
    return await invoke("create_ecosystem", { name, description, frameworks });
  },

  async switch(id: string): Promise<void> {
    return await invoke("switch_ecosystem", { id });
  },

  async delete(id: string): Promise<void> {
    return await invoke("delete_ecosystem", { id });
  },

  async listFrameworks(): Promise<FrameworkRegistry[]> {
    return await invoke("list_frameworks");
  },

  async installFramework(ecoId: string, frameworkId: string): Promise<void> {
    return await invoke("install_framework_to_ecosystem", {
      ecoId,
      frameworkId,
    });
  },

  async uninstallFramework(ecoId: string, frameworkId: string): Promise<void> {
    return await invoke("uninstall_framework_from_ecosystem", {
      ecoId,
      frameworkId,
    });
  },

  async updateFramework(ecoId: string, frameworkId: string): Promise<void> {
    return await invoke("update_framework_in_ecosystem", {
      ecoId,
      frameworkId,
    });
  },

  async getEcosystemFrameworks(ecoId: string): Promise<string[]> {
    return await invoke("get_ecosystem_frameworks", { ecoId });
  },

  async getEcosystemStatus(ecoId: string): Promise<EcosystemStatus> {
    return await invoke("get_ecosystem_status", { ecoId });
  },

  async saveUserPreferences(ecoId: string, fileName: string): Promise<void> {
    return await invoke("save_user_preferences", { ecoId, fileName });
  },

  async removeUserPreference(
    ecoId: string,
    fileName: string,
    keyPath: string,
  ): Promise<void> {
    return await invoke("remove_user_preference", { ecoId, fileName, keyPath });
  },
};

File: src-tauri/src/commands/ecosystem.rs
use crate::services::ecosystem::EcosystemService;
use crate::services::ecosystem_framework;
use crate::store::AppState;
use tauri::State;

#[tauri::command]
pub async fn create_ecosystem(
    state: State<'_, AppState>,
    name: String,
    description: String,
    frameworks: Vec<String>,
) -> Result<crate::database::Ecosystem, String> {
    EcosystemService::create(&state, &name, &description, frameworks).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn switch_ecosystem(state: State<'_, AppState>, id: String) -> Result<(), String> {
    EcosystemService::switch(&state, &id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_ecosystem(state: State<'_, AppState>, id: String) -> Result<(), String> {
    EcosystemService::delete(&state, &id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_ecosystems(
    state: State<'_, AppState>,
) -> Result<Vec<crate::database::Ecosystem>, String> {
    EcosystemService::list(&state).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_current_ecosystem(
    state: State<'_, AppState>,
) -> Result<Option<crate::database::Ecosystem>, String> {
    EcosystemService::get_current(&state).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_frameworks() -> Result<Vec<ecosystem_framework::FrameworkRegistry>, String> {
    Ok(ecosystem_framework::get_all_frameworks())
}

#[tauri::command]
pub async fn install_framework_to_ecosystem(
    state: State<'_, AppState>,
    eco_id: String,
    framework_id: String,
) -> Result<(), String> {
    EcosystemService::install_framework(&state, &eco_id, &framework_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn uninstall_framework_from_ecosystem(
    state: State<'_, AppState>,
    eco_id: String,
    framework_id: String,
) -> Result<(), String> {
    EcosystemService::uninstall_framework(&state, &eco_id, &framework_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_framework_in_ecosystem(
    state: State<'_, AppState>,
    eco_id: String,
    framework_id: String,
) -> Result<(), String> {
    EcosystemService::update_framework(&state, &eco_id, &framework_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_ecosystem_frameworks(eco_id: String) -> Result<Vec<String>, String> {
    EcosystemService::get_ecosystem_frameworks(&eco_id).map_err(|e| e.to_string())
}
